File: firebase/functions/lib/events/calendar/get_calendar_link.dart
import 'dart:async';

import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import '../../on_call_function.dart';
import '../../utils/calendar_link_util.dart';
import '../../utils/infra/firebase_auth_utils.dart';
import '../../utils/infra/firestore_utils.dart';
import '../../utils/utils.dart';
import 'package:data_models/cloud_functions/requests.dart';
import 'package:data_models/events/event.dart';
import 'package:data_models/community/community.dart';

import '../../utils/template_utils.dart';

class GetCommunityCalendarLink
    extends OnCallMethod<GetCommunityCalendarLinkRequest> {
  GetCommunityCalendarLink()
      : super(
          GetCommunityCalendarLinkRequest.functionName,
          (jsonMap) => GetCommunityCalendarLinkRequest.fromJson(jsonMap),
        );

  @override
  Future<Map<String, dynamic>> action(
    GetCommunityCalendarLinkRequest request,
    CallableContext context,
  ) async {
    orElseUnauthorized(
      context.authUid != null,
      logMessage: 'Context auth ID was null',
    );

    final event = await firestoreUtils.getFirestoreObject(
      path: request.eventPath,
      constructor: (map) => Event.fromJson(map),
    );
    final users = await firebaseAuthUtils.getUsers([event.creatorId]);
    if (users.length != 1) {
      throw Exception('Organizer not found');
    }
    final organizer = users[0];
    final communitySnapshot =
        await firestore.document('community/${event.communityId}').get();
    final community = Community.fromJson(
      firestoreUtils.fromFirestoreJson(communitySnapshot.data.toMap()),
    );

    final templatePath =
        'community/${community.id}/templates/${event.templateId}';
    final templateDoc = await firestore.document(templatePath).get();
    final template = TemplateUtils.templateFromSnapshot(templateDoc);

    final googleCalendarLink = calendarLinkUtil.getGoogleLink(
      community: community,
      template: template,
      event: event,
    );

    final office365CalendarLink = calendarLinkUtil.getOffice365Link(
      community: community,
      template: template,
      event: event,
    );

    final outlookCalendarLink = calendarLinkUtil.getOutlookLink(
      community: community,
      template: template,
      event: event,
    );

    final icsLink = calendarLinkUtil.getICS(
      community: community,
      template: template,
      event: event,
      organizer: organizer,
    );
    return GetCommunityCalendarLinkResponse(
      googleCalendarLink: googleCalendarLink,
      office365CalendarLink: office365CalendarLink,
      outlookCalendarLink: outlookCalendarLink,
      icsLink: icsLink,
    ).toJson();
  }
}

File: firebase/functions/src/events/calendar/get_calendar_link.ts
import * as functions from 'firebase-functions';
import { OnCallMethod } from '../../on_call_function';
import { firestore, firestoreUtils } from '../../utils/infra/firestore_utils';
import { firebaseAuthUtils } from '../../utils/infra/firebase_auth_utils';
import { calendarLinkUtil } from '../../utils/calendar_link_util';
import { TemplateUtils } from '../../utils/template_utils';
import { orElseUnauthorized } from '../../utils/utils';
import { Event } from '../../types';

interface GetCommunityCalendarLinkRequest {
  eventPath: string;
}

interface GetCommunityCalendarLinkResponse {
  googleCalendarLink?: string;
  office365CalendarLink?: string;
  outlookCalendarLink?: string;
  icsLink?: string;
}

export class GetCommunityCalendarLink extends OnCallMethod<GetCommunityCalendarLinkRequest> {
  constructor() {
    super('GetCommunityCalendarLink', (jsonMap) => jsonMap as GetCommunityCalendarLinkRequest);
  }

  async action(
    request: GetCommunityCalendarLinkRequest,
    context: functions.https.CallableContext
  ): Promise<GetCommunityCalendarLinkResponse> {
    orElseUnauthorized(context.auth?.uid != null, { logMessage: 'Context auth ID was null' });

    const event: Event = await firestoreUtils.getFirestoreObject({
      path: request.eventPath,
      constructor: (map) => map as unknown as Event,
    });

    const users = await firebaseAuthUtils.getUsers([(event as any).creatorId]);
    if (users.length !== 1) throw new Error('Organizer not found');
    const organizer = users[0];

    const communitySnap = await firestore.doc(`community/${(event as any).communityId}`).get();
    const community = firestoreUtils.fromFirestoreJson(communitySnap.data() ?? {}) as unknown as any;

    const templatePath = `community/${community.id}/templates/${(event as any).templateId}`;
    const templateDoc = await firestore.doc(templatePath).get();
    const template = TemplateUtils.templateFromSnapshot(templateDoc);

    return {
      googleCalendarLink: calendarLinkUtil.getGoogleLink({ community, template, event }),
      office365CalendarLink: calendarLinkUtil.getOffice365Link({ community, template, event }),
      outlookCalendarLink: calendarLinkUtil.getOutlookLink({ community, template, event }),
      icsLink: calendarLinkUtil.getICS({ community, template, event, organizerEmail: (organizer as any)?.email }),
    };
  }
}
